Improve visualization branch + Feature comparing

kntrain opened the pull request.

gliptak committed:
Bring Python to 3.9 in Travis

JohannesEbke committed:
Merge pull request #44 from gliptak/patch-2

Bring Python to 3.9 in Travis. Thanks!

kntrain committed:
-boto3 update and recreate caches

kntrain committed:
-updated boto3

JohannesEbke committed:
Merge pull request #46 from kntrain/boto3-update-1.16.21

Boto3 update 1.16.24

kntrain committed:
-added additional show options
-replaced run_raw call with file reference in wrapper

kntrain committed:
-moved modifications in resources to filter class

kntrain committed:
-fixed restructuring issues
-moved resources execution to apply_filter
-added command line option for resource filters

kntrain committed:
-removed call to file-reference filter in listing

JohannesEbke committed:
recreate_caches: Only accept "Host not found" as error, also Warn if run without credentials, as it will run very slowly.

JohannesEbke committed:
Upgrade to boto3 1.16.57

JohannesEbke committed:
Add heuristic for new host prefixes and recreate caches

JohannesEbke committed:
Fix up tests

JohannesEbke committed:
Yapf

kntrain committed:
-renamed Listing -> RawListing
-renamed ListingFile -> FilteredListing
-moved non-default filters to fixing_filter.py

kntrain committed:
changed fixing_filters to always apply

kntrain committed:
- new feature branch for visual improvements
- added print-html from old branch
- added count number

kntrain committed:
- added html search feature
- moved most html-printing functionality to new py-file

kntrain committed:
- fixed changing column width

kntrain committed:
- fixed dynamically changing count number

kntrain committed:
- upgraded html style
- fixed html-file call from directory

kntrain committed:
- added finish time to html

kntrain committed:
- added handling in case special ops are specified

kntrain committed:
- removed empty line

kntrain committed:
- added unfilter options description

kntrain committed:
Create HTML header and footer

- Combine title and searchbox in header
- Move timestamps to footer

kntrain committed:
Add comparing feature to available commands

kntrain committed:
Fix dependent ops and show command flags Bugs

kntrain committed:
Implement compare command between 2 directories

kntrain committed:
Move html related code to generate-html.py

kntrain committed:
Restructure to get resource ID during query

kntrain committed:
Remove stray comment and unfilter-list conversion

kntrain committed:
Add error-handling to JSON-file dependent function

kntrain committed:
Adopt suggestions from errorhandle-branch

-Remove stray comments
-Remove convert_unfilterList
-Add error-hadling to JSON-dependent resources-block
-Adjust resources-dependent query functions

kntrain committed:
Fix wrong do_list_files behaviour for errors

kntrain committed:
Add ResultListing as return Obj of acquire_listing

kntrain committed:
Enhance visualization-feature

-Finish display of comparison in HTML
-Add display of resource IDs for found resources
-Change acquire_listing return value to ResultListing object
-Refactor do_list_files into info summary and resource ID search

kntrain committed:
Fix comparing algorithm bug

-Fix comparing ResultListing-functions
-Add JSON-functions to FIlteredListing
-Save dir + unfilter in JSON-files
-Delete stray comments

JohannesEbke committed:
Merge pull request #49 from kntrain/improve-errorhandle

This list of patches splits out the raw listing from the filtered listings, and should allow better handling of different kinds of errors in the long term. In the short term, it adds the possibility to "unfilter" the automatically applied filters that until now were baked in.

JohannesEbke commented on aws_list_all/__main__.py:
```diff
@@ -119,6 +137,47 @@ def main():
+    viewhtml.add_argument('-p', '--parallel', default=32, type=int, help='Number of request to do in parallel')
+    viewhtml.add_argument('-d', '--directory', default='.', help='Directory to save result listings to')
+    viewhtml.add_argument('-v', '--verbose', action='count', help='Print detailed info during run')
+    viewhtml.add_argument('-c', '--profile', help='Use a specific .aws/credentials profile.')
```
Shouldn't viewhtml be more similar to "show"? This looks like it's doing all the querying again.

JohannesEbke commented on aws_list_all/__main__.py:
```diff
@@ -183,6 +250,45 @@ def main():
+
+        new = 2
+        url = os.getcwd() + "/compare.html"
+        webbrowser.open(url,new=new)
```
This section should go into its own module

JohannesEbke commented on aws_list_all/generate_html.py:
```diff
@@ -0,0 +1,225 @@
+    print('  margin-bottom: 12px;\n')
+    print('}\n')
+    print('</style>\n')
+    print('</head>\n')
```
The static stuff should probably go into a CONSTANT with triple quotes """, then it can be multiline?

JohannesEbke commented on aws_list_all/generate_html.py:
```diff
@@ -0,0 +1,225 @@
+            print('        </td>')
+        print('    </tr>\n')
+    print('</table>')
+    print('</div>')
```
I am tempted to suggest a more generic solution here, but this is actually quite readable.

JohannesEbke commented on aws_list_all/resource_filter.py:
```diff
@@ -0,0 +1,285 @@
+        if listing.service == 'elasticache' and listing.operation == 'DescribeCacheSubnetGroups':
+            response['CacheSubnetGroups'] = [
+                g for g in response.get('CacheSubnetGroups', []) if g.get('CacheSubnetGroupName') != 'default'
+            ]
```
This branch probably needs a rebase to the current master since I merged the other PR

JohannesEbke reviewed (changes requested):
Thanks, looking almost ready!

kntrain committed:
Do minor restrcuturing in main and generate_html

-Move access to html-file from main to generate_html
-Refactor sections from html-head into string constants

kntrain committed:
Incorporate print-html into new command 'view'

-Delete old print-html command
-Set html as option to 'view'
-Implement view's function to execute query+show or print html results

kntrain committed:
Incorporate html+compare option into show command

kntrain committed:
Fix html-file save location

JohannesEbke commented on aws_list_all/__main__.py:
```diff
@@ -183,6 +253,31 @@ def main():
+            os.chdir(args.directory)
+        increase_limit_nofiles()
+        services = args.service or get_services()
+        do_consecutive(
```
I think this could be expressed as a sequence of query & view.

JohannesEbke commented on aws_list_all/__main__.py:
```diff
@@ -119,6 +138,45 @@ def main():
+    # Combine the features of query and show and optionally print the findings
+    # from query to an HTML file to be viewed in a browser
+    view = subparsers.add_parser(
+        'view', description='Print a listing to an HTML file to be viewed in a browser',
```
This should probably be clearer about that it is actually querying the data.

JohannesEbke commented on aws_list_all/generate_html.py:
```diff
@@ -0,0 +1,271 @@
+    """Open a file to write HTML-content in and return the original system output and file path"""
+    origout = sys.stdout
+    f = open(name, 'w')
+    sys.stdout = f
```
I really don't like bending sys.stdout here like this. Ideally, all these functions should just add a string together (lines = []; lines.append instead of print; return "\n".join(lines))

JohannesEbke commented on aws_list_all/listing.py:
```diff
@@ -380,3 +275,59 @@ def resources(self):  # pylint:disable=too-many-branches
+    def diffInListing(cls, base, diff):
+        return cls(base.input, base.result_type, base.details, base.id_list, diff)
+
+    def __eq__(self, other):
```
Here you can just define __eq__ and __lt__ and then use https://docs.python.org/3/library/functools.html#functools.total_ordering to generate the rest

JohannesEbke reviewed (commented):
Looks good to me, modulo the issues raised. I'd probably merge and fix up as is, but feel free to update the PR!

kntrain committed:
Change creation and generating of html-content

-Change printing of html-strings to returning string at the end
-Write to html-file after finished joining all strings at the end
-Adjust view-command description and help

JohannesEbke commented:
OK, looks good for now. There are still some merge conflicts now?

kntrain committed:
Merge branch 'master' into feature_comparing
